# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/thebes/gfxTextRun.h
# Commit: f8ca02252dfc
# Full Hash: f8ca02252dfc48d7342703e6a635d510b9dc0b5d
# Author: Jonathan Kew <jkew@mozilla.com>
# Date: 2020-04-02 09:51:45
# Regressor Bug: 1619349
# File Overlap Count: 1
# Description:
#   Bug 1619349 - patch 1 - Record statistics about font-matching behavior in the presContext. r=jwatt
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D67787
# ==============================================================================

diff -r 01a2a527aa1a -r f8ca02252dfc gfx/thebes/gfxTextRun.h
--- a/gfx/thebes/gfxTextRun.h	Wed Apr 01 21:42:11 2020 +0000
+++ b/gfx/thebes/gfxTextRun.h	Wed Apr 01 21:42:36 2020 +0000
@@ -651,8 +651,6 @@
    */
   void FetchGlyphExtents(DrawTarget* aRefDrawTarget);
 
-  uint32_t CountMissingGlyphs() const;
-
   const GlyphRun* GetGlyphRuns(uint32_t* aNumGlyphRuns) const {
     if (mHasGlyphRunArray) {
       *aNumGlyphRuns = mGlyphRunArray.Length();
@@ -892,6 +890,36 @@
   ShapingState mShapingState;
 };
 
+enum class FallbackTypes : uint8_t {
+  // Font fallback used a font configured in Preferences
+  FallbackToPrefsFont = 1 << 0,
+  // Font fallback used a font with FontVisibility::Base
+  FallbackToBaseFont = 1 << 1,
+  // Font fallback used a font with FontVisibility::LangPack
+  FallbackToLangPackFont = 1 << 2,
+  // Font fallback used a font with FontVisibility::User
+  FallbackToUserFont = 1 << 3,
+  // Rendered missing-glyph because no font available for the character
+  MissingFont = 1 << 4,
+  // Rendered missing-glyph but a LangPack font could have been used
+  MissingFontLangPack = 1 << 5,
+  // Rendered missing-glyph but a User font could have been used
+  MissingFontUser = 1 << 6,
+};
+
+MOZ_MAKE_ENUM_CLASS_BITWISE_OPERATORS(FallbackTypes)
+
+struct FontMatchingStats {
+  // Set of names that have been looked up (whether successfully or not).
+  nsTHashtable<nsCStringHashKey> mFamilyNames;
+  // Number of font-family names resolved at each level of visibility.
+  uint32_t mBaseFonts = 0;
+  uint32_t mLangPackFonts = 0;
+  uint32_t mUserFonts = 0;
+  uint32_t mWebFonts = 0;
+  FallbackTypes mFallbacks = FallbackTypes(0);
+};
+
 class gfxFontGroup final : public gfxTextRunFactory {
  public:
   typedef mozilla::unicode::Script Script;
@@ -902,10 +930,13 @@
 
   gfxFontGroup(const mozilla::FontFamilyList& aFontFamilyList,
                const gfxFontStyle* aStyle, gfxTextPerfMetrics* aTextPerf,
+               FontMatchingStats* aFontMatchingStats,
                gfxUserFontSet* aUserFontSet, gfxFloat aDevToCssSize);
 
   virtual ~gfxFontGroup();
 
+  gfxFontGroup(const gfxFontGroup& aOther) = delete;
+
   // Returns first valid font in the fontlist or default font.
   // Initiates userfont loads if userfont not loaded.
   // aGeneric: if non-null, returns the CSS generic type that was mapped to
@@ -920,8 +951,6 @@
 
   const gfxFontStyle* GetStyle() const { return &mStyle; }
 
-  gfxFontGroup* Copy(const gfxFontStyle* aStyle);
-
   /**
    * The listed characters should be treated as invisible and zero-width
    * when creating textruns.
@@ -1352,6 +1381,8 @@
 
   gfxTextPerfMetrics* mTextPerf;
 
+  FontMatchingStats* mFontMatchingStats;
+
   // Cache a textrun representing an ellipsis (useful for CSS text-overflow)
   // at a specific appUnitsPerDevPixel size and orientation
   RefPtr<gfxTextRun> mCachedEllipsisTextRun;